crates/language-server/src/functionality/handlers.rs

```diff
@@ -0,0 +1,137 @@
+            }),
+        };
+        let _ = responder.send(Ok(initialize_result));
+    }
+
+    pub(super) async fn handle_shutdown(
+        &mut self,
+        responder: tokio::sync::oneshot::Sender<Result<(), tower_lsp::jsonrpc::Error>>,
+    ) {
+        info!("shutting down language server");
+        let _ = responder.send(Ok(()));
+    }
+
+    pub(super) async fn handle_deleted(
+        &mut self,
+        params: lsp_types::FileEvent,
+        tx_needs_diagnostics: tokio::sync::mpsc::UnboundedSender<String>,
+    ) {
+        let path = params.uri.to_file_path().unwrap();
+        info!("file deleted: {:?}", path);
+        let path = path.to_str().unwrap();
+        let _ = self
+            .workspace
+            .remove_input_for_file_path(&mut self.db, path);
+        let _ = tx_needs_diagnostics.send(path.to_string());
+    }
+
+    pub(super) async fn handle_change(
+        &mut self,
+        doc: TextDocumentItem,
+        tx_needs_diagnostics: tokio::sync::mpsc::UnboundedSender<String>,
+    ) {
+        info!("change detected: {:?}", doc.uri);
+        let path_buf = doc.uri.to_file_path().unwrap();
+        let path = path_buf.to_str().unwrap();
+        let contents = Some(doc.text);
+        if let Some(contents) = contents {
+            let input = self
+                .workspace
+                .touch_input_for_file_path(&mut self.db, path)
+                .unwrap();
+            let _ = input.sync_from_text(&mut self.db, contents);
+        }
+        let _ = tx_needs_diagnostics.send(path.to_string());
+    }
+
+    pub(super) async fn handle_diagnostics(&mut self, files_need_diagnostics: Vec<String>) {
+        let ingot_files_need_diagnostics: FxHashSet<_> = files_need_diagnostics
+            .into_iter()
+            .filter_map(|file| self.workspace.get_ingot_for_file_path(&file))
+            .flat_map(|ingot| ingot.files(self.db.as_input_db()))
+            .cloned()
+            .collect();
+
+        let db = self.db.snapshot();
+        let client = self.client.clone();
+        let compute_and_send_diagnostics = self
+            .workers
+            .spawn_blocking(move || {
+                db.get_lsp_diagnostics(ingot_files_need_diagnostics.into_iter().collect())
+            })
+            .and_then(|diagnostics| async move {
+                futures::future::join_all(diagnostics.into_iter().map(|(path, diagnostic)| {
+                    let client = client.clone();
+                    async move { client.publish_diagnostics(path, diagnostic, None).await }
+                }))
+                .await;
+                Ok(())
+            });
+        tokio::spawn(compute_and_send_diagnostics);
+    }
+
+    pub(super) async fn handle_hover(
+        &mut self,
+        params: lsp_types::HoverParams,
+        responder: tokio::sync::oneshot::Sender<
+            Result<Option<lsp_types::Hover>, tower_lsp::jsonrpc::Error>,
+        >,
+    ) {
+        let db = self.db.snapshot();
```

Comment:
There's no reason to use a snapshot here, right? `hover_handler(&self.db, ...` should suffice (unless this is changed so that the task is spawned off to run separately)

Reply:
Yup, this is also vestigial, I'd been handling hover in a worker but took it out and forgot to remove this.